pyam/core.py

```diff
@@ -733,118 +727,140 @@ def check_aggregate(self, variable, components=None, units=None,
-        df_region, df_components = (
-            _aggregate_by_regions(var_df.data, region, units)
-            .align(_aggregate_by_regions(var_df.data, components, units))
+        rows = self._apply_filters(region=region, variable=variable, unit=unit)
+        df_region, df_subregions = (
+            _aggregate(self.data[rows], ['region', 'variable'])
+            .align(df_subregions)
         )
 
-        df_components.index = df_components.index.droplevel(
-            "variable"
-        )
-
-        # Add in variables that are included in region totals but which
-        # aren't included in the regional components.
-        # For example, if we are looking at World and Emissions|BC, we need
-        # to add aviation and shipping to the sum of Emissions|BC for each
-        # of World's regional components to do a valid check.
-        different_region = components[0]
-        var_list = pd.Series(self.data.variable.unique())
-        var_components = var_list[pattern_match(var_list,
-                                                '{}|*'.format(variable), 0)]
-        for var_to_add in var_components:
-            var_rows = self.data.variable == var_to_add
-            region_rows = self.data.region == different_region
-            var_has_regional_info = (var_rows & region_rows).any()
-            if not var_has_regional_info:
-                df_var_to_add = self.filter(
-                    region=region, variable=var_to_add
-                ).data.groupby(REGION_IDX + ['unit']).sum()['value']
-
-                df_var_to_add.index = df_var_to_add.index.droplevel('variable')
-
-                if len(df_var_to_add):
-                    df_components = df_components.add(df_var_to_add,
-                                                      fill_value=0)
-
-        df_components = pd.concat([df_components], keys=[variable],
-                                  names=['variable'])
-
         # use `np.isclose` for checking match
-        diff = df_region[~np.isclose(df_region, df_components, **kwargs)]
+        diff = df_region[~np.isclose(df_region, df_subregions, **kwargs)]
 
         if len(diff):
             msg = (
-                '{} - {} of {} data points are not aggregates of regional '
-                'components'
+                '`{}` - {} of {} rows are not aggregates of subregions'
             )
             logger().info(msg.format(variable, len(diff), len(df_region)))
 
             if exclude_on_fail:
                 self._exclude_on_fail(diff.index.droplevel([2, 3]))
 
-            diff = pd.concat([diff], keys=[region], names=['region'])
-            diff.index = diff.index.swaplevel(i=-1, j=-2)
-            return diff.unstack().rename_axis(None, axis=1)
+            col_args = dict(region=region, variable=variable)
+            return IamDataFrame(diff, **col_args).timeseries()
+
+    def _variable_components(self, variable):
+        """Get all components of a variable"""
```

Comment:
I think it maybe worth showing an example in the docstring

Reply:
done